Failed to build Python bindings on Ubuntu with 3.x

Issue overview
--------------

reported on #3940 by @damianCaceres

## Current Behavior

When building the python bindings `-DBUILD_PYTHON_BINDINGS=ON`, user hit a build error.

```
damian@damian-HP:~$ conan --version
Conan version 1.24.1
```


```
In file included from /usr/include/python3.6m/Python.h:39:0,
                 from /home/damian/OpenStudio-3.0.0/build/src/radiance/python_OpenStudioRadiance_wrap.cxx:184:
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:35:53: error: ‘z_crc_t’ does not name a type
 static int decrypt_byte(unsigned long* pkeys, const z_crc_t* pcrc_32_tab)
                                                     ^~~~~~~
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:48:51: error: ‘z_crc_t’ does not name a type
 static int update_keys(unsigned long* pkeys,const z_crc_t* pcrc_32_tab,int c)
                                                   ^~~~~~~
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:65:69: error: ‘z_crc_t’ does not name a type
 static void init_keys(const char* passwd,unsigned long* pkeys,const z_crc_t* pcrc_32_tab)
                                                                     ^~~~~~~
src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/build.make:1374: recipe for target 'src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/python_OpenStudioRadiance_wrap.cxx.o' failed
make[2]: *** [src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/python_OpenStudioRadiance_wrap.cxx.o] Error 1
CMakeFiles/Makefile2:3627: recipe for target 'src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/all' failed
make[1]: *** [src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/all] Error 2
make[1]: *** Se espera a que terminen otras tareas....


In file included from /usr/include/python3.6m/Python.h:39:0,
                 from /home/damian/OpenStudio-3.0.0/build/src/gbxml/python_OpenStudioGBXML_wrap.cxx:184:
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:35:53: error: ‘z_crc_t’ does not name a type
 static int decrypt_byte(unsigned long* pkeys, const z_crc_t* pcrc_32_tab)
                                                     ^~~~~~~
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:48:51: error: ‘z_crc_t’ does not name a type
 static int update_keys(unsigned long* pkeys,const z_crc_t* pcrc_32_tab,int c)
                                                   ^~~~~~~
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:65:69: error: ‘z_crc_t’ does not name a type
 static void init_keys(const char* passwd,unsigned long* pkeys,const z_crc_t* pcrc_32_tab)
                                                                     ^~~~~~~
src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/build.make:1373: recipe for target 'src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/python_OpenStudioGBXML_wrap.cxx.o' failed
make[2]: *** [src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/python_OpenStudioGBXML_wrap.cxx.o] Error 1
CMakeFiles/Makefile2:3775: recipe for target 'src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/all' failed
make[1]: *** [src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/all] Error 2


```

User reported trying to change
https://github.com/NREL/OpenStudio/blob/17912d4d6be39e999fdd174908189b8a61b2ada5/ConanInstall.cmake#L75

to `zlib/1.2.11@conan/stable`: 

> Yes. I've changed the line 75 in the file ConanInstall.cmake
I know because OpenStudio SDK v3.0.0-pre1 works fine
Thank you very much

>I can compile with zlib/1.2.11@conan/stable, but the bindings have problems when they are imported. Now I'm starting to test them.
I'll come back to OpenStudio 2.9.1, which works very well. NO URGENT
Thank you

> After compile OpenStudio 3.0.0 I have got
> ```
> /usr/local/OpenStudio-3.0.0/Python/_openstudiomodel.so: undefined symbol: _ZTIN10openstudio9WorkspaceE
> ```
> I imagine that is due to swig 4 ... (wrong comment)

## Expected Behavior

It should work to build the Python bindings

## Steps to Reproduce

Use Ubuntu 18.04, build with  `-DBUILD_PYTHON_BINDINGS=ON`


## Details

### Environment

Some additional details about your environment for this issue (if relevant):
 - Platform (Operating system, version): Ubuntu 18.04
 - Version of OpenStudio (if using an intermediate build, include SHA): 3.0.0, 17912d4d6be39e999fdd174908189b8a61b2ada5

### Context

Trying to build the Python Bindings

Firstly WITH PYTHON2 BINDINGS

```

damian@damian-hp:~/OpenStudio-3.0.0-python2.7/build$ cmake .. -DBUILD_PYTHON_BINDINGS=ON -DPYTHON_EXECUTABLE=/usr/bin/python2.7 -DPYTHON_INCLUDE_DIR=/usr/include/python2.7 -DPYTHON_LIBRARY=/usr/lib/x86_64-linux-gnu/libpython2.7.so -DCMAKE_INSTALL_PREFIX=/usr/local/openstudio-3.0.0-python2.7
-- Configuring done
-- Generating done
-- Build files have been written to: /home/damian/OpenStudio-3.0.0-python2.7/build
damian@damian-hp:~/OpenStudio-3.0.0-python2.7/build$ make -j6
[ 25%] Built target openstudio_model
Makefile:171: recipe for target 'all' failed
make: *** [all] Error 2

```

```

[  8%] Building CXX object src/energyplus/CMakeFiles/openstudio_energyplus.dir/ForwardTranslator/ForwardTranslateConstruction.cpp.o
In file included from /usr/include/python2.7/Python.h:47:0,
                 from /home/damian/OpenStudio-3.0.0-python2.7/build/src/radiance/python_OpenStudioRadiance_wrap.cxx:184:
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:35:53: error: ‘z_crc_t’ does not name a type
 static int decrypt_byte(unsigned long* pkeys, const z_crc_t* pcrc_32_tab)
                                                     ^~~~~~~
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:48:51: error: ‘z_crc_t’ does not name a type
 static int update_keys(unsigned long* pkeys,const z_crc_t* pcrc_32_tab,int c)
                                                   ^~~~~~~
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:65:69: error: ‘z_crc_t’ does not name a type
 static void init_keys(const char* passwd,unsigned long* pkeys,const z_crc_t* pcrc_32_tab)
                                                                     ^~~~~~~
src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/build.make:1391: recipe for target 'src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/python_OpenStudioRadiance_wrap.cxx.o' failed
make[2]: *** [src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/python_OpenStudioRadiance_wrap.cxx.o] Error 1
CMakeFiles/Makefile2:2838: recipe for target 'src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/all' failed
make[1]: *** [src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[  8%] Building CXX object src/energyplus/CMakeFiles/openstudio_energyplus.dir/ForwardTranslator/ForwardTranslateConstructionAirBoundary.cpp.o

```

```

[ 11%] Building CXX object src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/python_OpenStudioGBXML_wrap.cxx.o
In file included from /usr/include/python2.7/Python.h:47:0,
                 from /home/damian/OpenStudio-3.0.0-python2.7/build/src/gbxml/python_OpenStudioGBXML_wrap.cxx:184:
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:35:53: error: ‘z_crc_t’ does not name a type
 static int decrypt_byte(unsigned long* pkeys, const z_crc_t* pcrc_32_tab)
                                                     ^~~~~~~
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:48:51: error: ‘z_crc_t’ does not name a type
 static int update_keys(unsigned long* pkeys,const z_crc_t* pcrc_32_tab,int c)
                                                   ^~~~~~~
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:65:69: error: ‘z_crc_t’ does not name a type
 static void init_keys(const char* passwd,unsigned long* pkeys,const z_crc_t* pcrc_32_tab)
                                                                     ^~~~~~~
[ 11%] Building CXX object src/utilities/CMakeFiles/openstudio_utilities.dir/idf/WorkspaceObjectDiff.cpp.o
[ 11%] Building CXX object src/utilities/CMakeFiles/openstudio_utilities.dir/idf/WorkspaceObjectWatcher.cpp.o
[ 11%] Building CXX object src/energyplus/CMakeFiles/openstudio_energyplus.dir/ForwardTranslator/ForwardTranslateFanConstantVolume.cpp.o
src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/build.make:1390: recipe for target 'src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/python_OpenStudioGBXML_wrap.cxx.o' failed
make[2]: *** [src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/python_OpenStudioGBXML_wrap.cxx.o] Error 1
CMakeFiles/Makefile2:2955: recipe for target 'src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/all' failed
make[1]: *** [src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/all] Error 2
[ 11%] Building CXX object src/energyplus/CMakeFiles/openstudio_energyplus.dir/ForwardTranslator/ForwardTranslateFanOnOff.cpp.o

```

```

damian@damian-hp:~/OpenStudio-3.0.0-python2.7/build$ grep "PYTHON" CMakeCache.txt
BUILD_PYTHON_BINDINGS:BOOL=ON
PYTHON_EXECUTABLE:FILEPATH=/usr/bin/python2.7
PYTHON_INCLUDE_DIR:PATH=/usr/include/python2.7
PYTHON_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libpython2.7.so
PYTHON_LIBRARY_DEBUG:FILEPATH=PYTHON_LIBRARY_DEBUG-NOTFOUND
//ADVANCED property for variable: PYTHON_EXECUTABLE
PYTHON_EXECUTABLE-ADVANCED:INTERNAL=1
//ADVANCED property for variable: PYTHON_INCLUDE_DIR
PYTHON_INCLUDE_DIR-ADVANCED:INTERNAL=1
//ADVANCED property for variable: PYTHON_LIBRARY
PYTHON_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: PYTHON_LIBRARY_DEBUG
PYTHON_LIBRARY_DEBUG-ADVANCED:INTERNAL=1
damian@damian-hp:~/OpenStudio-3.0.0-python2.7/build$ 

```

Secondly WITH PYTHON3 BINDINGS

```
damian@damian-hp:~/OpenStudio-3.0.0-python3.6/build$ cmake .. -DBUILD_PYTHON_BINDINGS=ON -DPYTHON_EXECUTABLE=/usr/bin/python3.6m -DPYTHON_INCLUDE_DIR=/usr/include/python3.6m -DPYTHON_LIBRARY=/usr/lib/x86_64-linux-gnu/libpython3.6m.so -DCMAKE_INSTALL_PREFIX=/usr/local/openstudio-3.0.0-python3.6
-- Configuring done
-- Generating done
-- Build files have been written to: /home/damian/OpenStudio-3.0.0-python3.6/build
damian@damian-hp:~/OpenStudio-3.0.0-python3.6/build$ make -j6
[ 25%] Built target openstudio_model
Makefile:171: recipe for target 'all' failed
make: *** [all] Error 2

```

```

[ 17%] Building CXX object src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/python_OpenStudioRadiance_wrap.cxx.o
In file included from /usr/include/python3.6m/Python.h:39:0,
                 from /home/damian/OpenStudio-3.0.0-python3.6/build/src/radiance/python_OpenStudioRadiance_wrap.cxx:184:
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:35:53: error: ‘z_crc_t’ does not name a type
 static int decrypt_byte(unsigned long* pkeys, const z_crc_t* pcrc_32_tab)
                                                     ^~~~~~~
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:48:51: error: ‘z_crc_t’ does not name a type
 static int update_keys(unsigned long* pkeys,const z_crc_t* pcrc_32_tab,int c)
                                                   ^~~~~~~
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:65:69: error: ‘z_crc_t’ does not name a type
 static void init_keys(const char* passwd,unsigned long* pkeys,const z_crc_t* pcrc_32_tab)
                                                                     ^~~~~~~
[ 17%] Building CXX object src/model/CMakeFiles/openstudio_model.dir/HotWaterEquipmentDefinition.cpp.o
[ 17%] Building CXX object src/energyplus/CMakeFiles/openstudio_energyplus.dir/ForwardTranslator/ForwardTranslateRefrigerationCondenserAirCooled.cpp.o
[ 17%] Building CXX object src/energyplus/CMakeFiles/openstudio_energyplus.dir/ForwardTranslator/ForwardTranslateRefrigerationCondenserCascade.cpp.o
src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/build.make:1391: recipe for target 'src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/python_OpenStudioRadiance_wrap.cxx.o' failed
make[2]: *** [src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/python_OpenStudioRadiance_wrap.cxx.o] Error 1
CMakeFiles/Makefile2:2838: recipe for target 'src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/all' failed
make[1]: *** [src/radiance/CMakeFiles/python_OpenStudioRadiance.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 17%] Building CXX object src/energyplus/CMakeFiles/openstudio_energyplus.dir/ForwardTranslator/ForwardTranslateRefrigerationCondenserEvaporativeCooled.cpp.o

```

```

[ 18%] Building CXX object src/model/CMakeFiles/openstudio_model.dir/OutputTableSummaryReports.cpp.o
In file included from /usr/include/python3.6m/Python.h:39:0,
                 from /home/damian/OpenStudio-3.0.0-python3.6/build/src/gbxml/python_OpenStudioGBXML_wrap.cxx:184:
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:35:53: error: ‘z_crc_t’ does not name a type
 static int decrypt_byte(unsigned long* pkeys, const z_crc_t* pcrc_32_tab)
                                                     ^~~~~~~
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:48:51: error: ‘z_crc_t’ does not name a type
 static int update_keys(unsigned long* pkeys,const z_crc_t* pcrc_32_tab,int c)
                                                   ^~~~~~~
/home/damian/.conan/data/zlib/1.2.11/_/_/package/aff0eb9e4c299cac0972a5fe39033141b3e50a6d/include/crypt.h:65:69: error: ‘z_crc_t’ does not name a type
 static void init_keys(const char* passwd,unsigned long* pkeys,const z_crc_t* pcrc_32_tab)
                                                                     ^~~~~~~
[ 18%] Building CXX object src/energyplus/CMakeFiles/openstudio_energyplus.dir/ForwardTranslator/ForwardTranslateThermalZone.cpp.o
src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/build.make:1390: recipe for target 'src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/python_OpenStudioGBXML_wrap.cxx.o' failed
make[2]: *** [src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/python_OpenStudioGBXML_wrap.cxx.o] Error 1
CMakeFiles/Makefile2:2955: recipe for target 'src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/all' failed
make[1]: *** [src/gbxml/CMakeFiles/python_OpenStudioGBXML.dir/all] Error 2
[ 18%] Building CXX object src/energyplus/CMakeFiles/openstudio_energyplus.dir/ForwardTranslator/ForwardTranslateThermostatSetpointDualSetpoint.cpp.o

```

```

damian@damian-hp:~/OpenStudio-3.0.0-python3.6/build$ grep "PYTHON" CMakeCache.txt
BUILD_PYTHON_BINDINGS:BOOL=ON
PYTHON_EXECUTABLE:FILEPATH=/usr/bin/python3.6m
PYTHON_INCLUDE_DIR:PATH=/usr/include/python3.6m
PYTHON_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libpython3.6m.so
PYTHON_LIBRARY_DEBUG:FILEPATH=PYTHON_LIBRARY_DEBUG-NOTFOUND
//ADVANCED property for variable: PYTHON_EXECUTABLE
PYTHON_EXECUTABLE-ADVANCED:INTERNAL=1
//ADVANCED property for variable: PYTHON_INCLUDE_DIR
PYTHON_INCLUDE_DIR-ADVANCED:INTERNAL=1
//ADVANCED property for variable: PYTHON_LIBRARY
PYTHON_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: PYTHON_LIBRARY_DEBUG
PYTHON_LIBRARY_DEBUG-ADVANCED:INTERNAL=1

```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -801,11 +801,10 @@ if(BUILD_NODE_MODULES)
 endif()
 
 if(BUILD_PYTHON_BINDINGS)
-  # need python
-  # TODO: add support for 3.x
-  find_package(PythonInterp 2.7 REQUIRED)
-  find_package(PythonLibs 2.7 REQUIRED)
-  include_directories(SYSTEM ${PYTHON_INCLUDE_DIRS})
+
+  find_package(Python COMPONENTS Interpreter Development REQUIRED)
+
+  include_directories(SYSTEM ${Python_INCLUDE_DIRS})
   set(ALL_PYTHON_BINDINGS "") # global list of python bindings
   set(ALL_PYTHON_BINDING_DEPENDS "") # global list of library dependencies of the generated wrapper cxx files
   set(ALL_PYTHON_WRAPPER_FILES "") # global list of generated wrapper cxx files

--- a/ConanInstall.cmake
+++ b/ConanInstall.cmake
@@ -72,7 +72,7 @@ if(NOT CONAN_OPENSTUDIO_ALREADY_RUN)
     #boost_geometry/${BOOST_VERSION}@bincrafters/stable
     pugixml/1.10@bincrafters/stable
     jsoncpp/1.9.2
-    zlib/1.2.11
+    zlib/1.2.11@nrel/testing # TODO: Temp, pending merging of https://github.com/conan-io/conan-center-index/pull/1526, to resolve #3961
     fmt/6.0.0
     sqlite3/3.30.1
     cpprestsdk/2.10.14@bincrafters/stable

--- a/ProjectMacros.cmake
+++ b/ProjectMacros.cmake
@@ -503,6 +503,7 @@ macro(MAKE_SWIG_TARGET NAME SIMPLENAME KEY_I_FILE I_FILES PARENT_TARGET PARENT_S
       ${SWIG_WRAPPER}
     )
 
+    # TODO: for local testing, PYTHON_GENERATED_SRC should go into Products/python next to the .so files
     install(FILES "${PYTHON_GENERATED_SRC}" DESTINATION Python COMPONENT "Python")
     install(TARGETS ${swig_target} DESTINATION Python COMPONENT "Python")
 

--- a/ProjectMacros.cmake
+++ b/ProjectMacros.cmake
@@ -435,7 +435,7 @@ macro(MAKE_SWIG_TARGET NAME SIMPLENAME KEY_I_FILE I_FILES PARENT_TARGET PARENT_S
   set(ALL_RUBY_BINDING_WRAPPERS_FULL_PATH "${ALL_RUBY_BINDING_WRAPPERS_FULL_PATH}" PARENT_SCOPE)
 
   # Python bindings
-  if(PYTHON_LIBRARY AND BUILD_PYTHON_BINDINGS)
+  if(BUILD_PYTHON_BINDINGS)
     set(swig_target "python_${NAME}")
 
     # utilities goes into OpenStudio. directly, everything else is nested
@@ -464,15 +464,17 @@ macro(MAKE_SWIG_TARGET NAME SIMPLENAME KEY_I_FILE I_FILES PARENT_TARGET PARENT_S
 
     # Add the -py3 flag if the version used is Python 3
     set(SWIG_PYTHON_3_FLAG "")
-    if (PYTHON_VERSION_MAJOR)
-      if (PYTHON_VERSION_MAJOR EQUAL 3)
+    if (Python_VERSION_MAJOR)
+      if (Python_VERSION_MAJOR EQUAL 3)
         set(SWIG_PYTHON_3_FLAG -py3)
         message(STATUS "${MODULE} - Building SWIG Bindings for Python 3")
       else()
         message(STATUS "${MODULE} - Building SWIG Bindings for Python 2")
       endif()
     else()
-      message(STATUS "${MODULE} - Couldnt determine version of Python - Building SWIG Bindings for Python 2")
+      # Python2 has been EOL since January 1, 2020
+      set(SWIG_PYTHON_3_FLAG -py3)
+      message(STATUS "${MODULE} - Couldnt determine version of Python - Building SWIG Bindings for Python 3")
     endif()
 
     add_custom_command(
@@ -520,7 +522,7 @@ macro(MAKE_SWIG_TARGET NAME SIMPLENAME KEY_I_FILE I_FILES PARENT_TARGET PARENT_S
       endif()
     endif()
 
-    target_link_libraries(${swig_target} ${PARENT_TARGET} ${PYTHON_LIBRARY})
+    target_link_libraries(${swig_target} ${PARENT_TARGET} ${PYTHON_Libraries})
 
     add_dependencies(${swig_target} ${PARENT_TARGET})
 

--- a/python/CMakeLists.txt
+++ b/python/CMakeLists.txt
@@ -21,7 +21,7 @@
 #target_link_libraries(openstudio_python
 # ${ALL_PYTHON_BINDING_DEPENDS}
 # ${Boost_LIBRARIES}
-# ${PYTHON_LIBRARY}
+# ${Python_LIBRARIES}
 #)
 
 #install(TARGETS openstudio_python DESTINATION Python CONFIGURATIONS COMPONENT "Python")

--- a/python/README.txt
+++ b/python/README.txt
@@ -1,7 +1,11 @@
 Example for launching example
 
+Note: the *.py files are generated in <build>/python_wrapper/generated_sources while the library files are in <build>/Products/python.
+When installing, they will end up in the same directory. You can either go symlink the generated_sources/*.py to Products/python, or specify both
+in PYTHONPATH
+
 On Linux:
-PYTHONPATH=~/openstudio/build/OSCore-prefix/src/OSCore-build/Products/python/ python testpath.py
+PYTHONPATH=~/openstudio/build/Products/python/:~/openstudio/build/python_wrapper/generated_sources/ python testpath.py
 
 On Windows:
 set PYTHONPATH=\working\openstudio\build\OSCore-prefix\src\OSCore-build\Products\python\Release

--- a/python/openstudio.py
+++ b/python/openstudio.py
@@ -1,11 +1,8 @@
 # imports all of the openstudio libraries into a friendly namespace
 import .openstudioairflow as airflow
-import .openstudioanalysis as analysis
-import .openstudioanalysisdriver as analysisdriver
 import .openstudioenergyplus as energyplus
 import .openstudiogbxml as gbxml
 import .openstudioisomodel as isomodel
-import .openstudiolib as lib
 import .openstudiomeasure as measure
 import .openstudiomodel as model
 # import .openstudiomodelcore as modelcore
@@ -18,12 +15,10 @@ import .openstudiomodel as model
 import .openstudioosversion as osversion
 import .openstudioradiance as radiance
 import .openstudiosdd as sdd
-# import .openstudiomodeleditor as modeleditor
 from .openstudioutilities import *
 from .openstudioutilitiesbcl import *
 from .openstudioutilitiescore import *
 from .openstudioutilitiesdata import *
-from .openstudioutilitieseconomics import *
 from .openstudioutilitiesgeometry import *
 from .openstudioutilitiesidd import *
 from .openstudioutilitiesidf import *

--- a/python/openstudio.py
+++ b/python/openstudio.py
@@ -1,34 +1,73 @@
 # imports all of the openstudio libraries into a friendly namespace
-import .openstudioairflow as airflow
-import .openstudioenergyplus as energyplus
-import .openstudiogbxml as gbxml
-import .openstudioisomodel as isomodel
-import .openstudiomeasure as measure
-import .openstudiomodel as model
-# import .openstudiomodelcore as modelcore
-# import .openstudiomodelgenerators as modelgenerators
-# import .openstudiomodelgeometry as modelgeometry
-# import .openstudiomodelhvac as modelhvac
-# import .openstudiomodelrefrigeration as modelrefrigeration
-# import .openstudiomodelresources as modelresources
-# import .openstudiomodelsimulation as modelsimulation
-# import .openstudiomodelairflow as modelairflow
-# import .openstudiomodelavailabilitymanager as modelavailabilitymanager
-# import .openstudiomodelplantequipmentoperationscheme as modelplantequipmentoperationscheme
-# import .openstudiomodelstraightcomponent as modelstraightcomponent
-# import .openstudiomodelzonehvac as momodelzonehvac
-import .openstudioosversion as osversion
-import .openstudioradiance as radiance
-import .openstudiosdd as sdd
-from .openstudioutilities import *
-from .openstudioutilitiesbcl import *
-from .openstudioutilitiescore import *
-from .openstudioutilitiesdata import *
-from .openstudioutilitiesgeometry import *
-from .openstudioutilitiesidd import *
-from .openstudioutilitiesidf import *
-# from .openstudioutilitiesfiletypes import *
-# from .openstudioutilitiesplot import *
-from .openstudioutilitiessql import *
-from .openstudioutilitiestime import *
-from .openstudioutilitiesunits import *
+
+if __package__ or "." in __name__:
+
+    from . import openstudioairflow as airflow
+    from . import openstudioenergyplus as energyplus
+    from . import openstudiogbxml as gbxml
+    from . import openstudioisomodel as isomodel
+    from . import openstudiomeasure as measure
+    from . import openstudiomodel as model
+    # from . import openstudiomodelcore as modelcore
+    # from . import openstudiomodelgenerators as modelgenerators
+    # from . import openstudiomodelgeometry as modelgeometry
+    # from . import openstudiomodelhvac as modelhvac
+    # from . import openstudiomodelrefrigeration as modelrefrigeration
+    # from . import openstudiomodelresources as modelresources
+    # from . import openstudiomodelsimulation as modelsimulation
+    # from . import openstudiomodelairflow as modelairflow
+    # from . import openstudiomodelavailabilitymanager as modelavailabilitymanager
+    # from . import openstudiomodelplantequipmentoperationscheme as modelplantequipmentoperationscheme
+    # from . import openstudiomodelstraightcomponent as modelstraightcomponent
+    # from . import openstudiomodelzonehvac as momodelzonehvac
+    from . import openstudioosversion as osversion
+    from . import openstudioradiance as radiance
+    from . import openstudiosdd as sdd
+    from .openstudioutilities import *
+    from .openstudioutilitiesbcl import *
+    from .openstudioutilitiescore import *
+    from .openstudioutilitiesdata import *
+    from .openstudioutilitiesgeometry import *
+    from .openstudioutilitiesidd import *
+    from .openstudioutilitiesidf import *
+    # from .openstudioutilitiesfiletypes import *
+    # from .openstudioutilitiesplot import *
+    from .openstudioutilitiessql import *
+    from .openstudioutilitiestime import *
+    from .openstudioutilitiesunits import *
+
+else:
+
+    import openstudioairflow as airflow
+    import openstudioenergyplus as energyplus
+    import openstudiogbxml as gbxml
+    import openstudioisomodel as isomodel
+    import openstudiomeasure as measure
+    import openstudiomodel as model
+    # import openstudiomodelcore as modelcore
+    # import openstudiomodelgenerators as modelgenerators
+    # import openstudiomodelgeometry as modelgeometry
+    # import openstudiomodelhvac as modelhvac
+    # import openstudiomodelrefrigeration as modelrefrigeration
+    # import openstudiomodelresources as modelresources
+    # import openstudiomodelsimulation as modelsimulation
+    # import openstudiomodelairflow as modelairflow
+    # import openstudiomodelavailabilitymanager as modelavailabilitymanager
+    # import openstudiomodelplantequipmentoperationscheme as modelplantequipmentoperationscheme
+    # import openstudiomodelstraightcomponent as modelstraightcomponent
+    # import openstudiomodelzonehvac as momodelzonehvac
+    import openstudioosversion as osversion
+    import openstudioradiance as radiance
+    import openstudiosdd as sdd
+    from openstudioutilities import *
+    from openstudioutilitiesbcl import *
+    from openstudioutilitiescore import *
+    from openstudioutilitiesdata import *
+    from openstudioutilitiesgeometry import *
+    from openstudioutilitiesidd import *
+    from openstudioutilitiesidf import *
+    # from openstudioutilitiesfiletypes import *
+    # from openstudioutilitiesplot import *
+    from openstudioutilitiessql import *
+    from openstudioutilitiestime import *
+    from openstudioutilitiesunits import *

--- a/python/openstudio.py
+++ b/python/openstudio.py
@@ -1,5 +1,4 @@
 # imports all of the openstudio libraries into a friendly namespace
-
 import .openstudioairflow as airflow
 import .openstudioanalysis as analysis
 import .openstudioanalysisdriver as analysisdriver
@@ -9,18 +8,17 @@ import .openstudioisomodel as isomodel
 import .openstudiolib as lib
 import .openstudiomeasure as measure
 import .openstudiomodel as model
-# import openstudiomodelcore as modelcore
-# import openstudiomodelgenerators as modelgenerators
-# import openstudiomodelgeometry as modelgeometry
-# import openstudiomodelhvac as modelhvac
-# import openstudiomodelrefrigeration as modelrefrigeration
-# import openstudiomodelresources as modelresources
-# import openstudiomodelsimulation as modelsimulation
+# import .openstudiomodelcore as modelcore
+# import .openstudiomodelgenerators as modelgenerators
+# import .openstudiomodelgeometry as modelgeometry
+# import .openstudiomodelhvac as modelhvac
+# import .openstudiomodelrefrigeration as modelrefrigeration
+# import .openstudiomodelresources as modelresources
+# import .openstudiomodelsimulation as modelsimulation
 import .openstudioosversion as osversion
 import .openstudioradiance as radiance
 import .openstudiosdd as sdd
 # import .openstudiomodeleditor as modeleditor
-
 from .openstudioutilities import *
 from .openstudioutilitiesbcl import *
 from .openstudioutilitiescore import *
@@ -34,5 +32,3 @@ from .openstudioutilitiesidf import *
 from .openstudioutilitiessql import *
 from .openstudioutilitiestime import *
 from .openstudioutilitiesunits import *
-
-

--- a/python/openstudio.py
+++ b/python/openstudio.py
@@ -12,6 +12,11 @@ import .openstudiomodel as model
 # import .openstudiomodelrefrigeration as modelrefrigeration
 # import .openstudiomodelresources as modelresources
 # import .openstudiomodelsimulation as modelsimulation
+# import .openstudiomodelairflow as modelairflow
+# import .openstudiomodelavailabilitymanager as modelavailabilitymanager
+# import .openstudiomodelplantequipmentoperationscheme as modelplantequipmentoperationscheme
+# import .openstudiomodelstraightcomponent as modelstraightcomponent
+# import .openstudiomodelzonehvac as momodelzonehvac
 import .openstudioosversion as osversion
 import .openstudioradiance as radiance
 import .openstudiosdd as sdd

--- a/python/openstudio_dynamic.py
+++ b/python/openstudio_dynamic.py
@@ -1,0 +1,69 @@
+# imports all of the openstudio libraries into a friendly namespace
+
+
+def import_module(module_name, namespace=None):
+    """ Helper to require a given module, and place it in a namespace or import
+    all in global namespace.
+
+    """
+    statement = None
+    if namespace:
+        if __package__ or "." in __name__:
+            statement = "from . import {} as {}".format(module_name, namespace)
+        else:
+            statement = "import {} as {}".format(module_name, namespace)
+    else:
+        if __package__ or "." in __name__:
+            statement = "from .{} import *".format(module_name)
+        else:
+            statement = "from {} import *".format(module_name)
+    try:
+        exec(statement)
+    except ImportError as e:
+        # Output expected (until fixed) ImportErrors.
+        print(e.__class__.__name__ + ": " + str(e))
+    except Exception as e:
+        # Output unexpected Exceptions.
+        print(e, False)
+        print(e.__class__.__name__ + ": " + str(e))
+
+
+imports = [
+    # (Module, Namespace)
+    ('openstudioairflow', 'airflow'),
+    ('openstudioenergyplus', 'energyplus'),
+    ('openstudiogbxml', 'gbxml'),
+    ('openstudioisomodel', 'isomodel'),
+    ('openstudiomeasure', 'measure'),
+    ('openstudiomodel', 'model'),
+    # ('openstudiomodelcore', 'modelcore'),
+    # ('openstudiomodelgenerators', 'modelgenerators'),
+    # ('openstudiomodelgeometry', 'modelgeometry'),
+    # ('openstudiomodelhvac', 'modelhvac'),
+    # ('openstudiomodelrefrigeration', 'modelrefrigeration'),
+    # ('openstudiomodelresources', 'modelresources'),
+    # ('openstudiomodelsimulation', 'modelsimulation'),
+    # ('openstudiomodelairflow', 'modelairflow'),
+    # ('openstudiomodelavailabilitymanager', 'modelavailabilitymanager'),
+    # ('openstudiomodelplantequipmentoperationscheme', 'modelplantequipmentoperationscheme'),
+    # ('openstudiomodelstraightcomponent', 'modelstraightcomponent'),
+    # ('openstudiomodelzonehvac', 'momodelzonehvac'),
+    ('openstudioosversion', 'osversion'),
+    ('openstudioradiance', 'radiance'),
+    ('openstudiosdd', 'sdd'),
+    ('openstudioutilities', None),
+    ('openstudioutilitiesbcl', None),
+    ('openstudioutilitiescore', None),
+    ('openstudioutilitiesdata', None),
+    ('openstudioutilitiesgeometry', None),
+    ('openstudioutilitiesidd', None),
+    ('openstudioutilitiesidf', None),
+    # ('openstudioutilitiesfiletypes', None),
+    # ('openstudioutilitiesplot', None),
+    ('openstudioutilitiessql', None),
+    ('openstudioutilitiestime', None),
+    ('openstudioutilitiesunits', None),
+]
+
+for (module_name, namespace) in imports:
+    import_module(module_name, namespace)

--- a/python/testpath.py
+++ b/python/testpath.py
@@ -1,15 +1,17 @@
 import openstudio
 
-print openstudio.toString(openstudio.runmanager.RunManager().dbPath())
+print(openstudio.toString(openstudio.runmanager.RunManager().dbPath()))
 
 # workspace test
 workspace = openstudio.Workspace()
-zone = workspace.addObject(openstudio.IdfObject(openstudio.IddObjectType("Zone")))
+zone = workspace.addObject(
+    openstudio.IdfObject(openstudio.IddObjectType("Zone"))
+)
 zone = zone.get()
 zone.setName("New Zone")
 
 for obj in workspace.objects():
-  print obj
+    print(obj)
 
 # model test
 model = openstudio.model.Model()
@@ -17,4 +19,4 @@ space = openstudio.model.Space(model)
 space.setName("New Space")
 
 for s in openstudio.model.getSpaces(model):
-  print s
+    print(s)

--- a/python/testpath.py
+++ b/python/testpath.py
@@ -1,4 +1,5 @@
-import openstudio
+# import openstudio
+import openstudio_dynamic as openstudio
 
 print(openstudio.toString(openstudio.runmanager.RunManager().dbPath()))
 

